### PR TITLE
[numpy] add op tri

### PR DESCRIPTION
**Pull request by Yiyan66**
## Description ##
add op tri 

  &#160;  | Old FFI (ctypes) (us) | New FFI (cython) (us)
:-------: | :-------------------: | :-: 
tri       | 37.32                 | 5.67

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/np_tri_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
+    std::ostringstream N_s, M_s, k_s, dtype_s;
+    N_s << N;
+    M_s << M;
+    k_s << k;
+    dtype_s << dtype;
+    (*dict)["N"] = N_s.str();
+    (*dict)["M"] = M_s.str();
+    (*dict)["k"] = k_s.str();
+    (*dict)["dtype"] = dtype_s.str();
```
Comment (hzfan) [inline src/operator/numpy/np_tri_op-inl.h]:
```
(*dict)["dtype"] = String2MXNetTypeWithBool(dtype);
```
`String2MXNetTypeWithBool` lies in `incubator-mxnet/src/api/operator/op_utils.h`

File: python/mxnet/ndarray/numpy/_op.py [hunk 2]
```diff
@@ -2015,6 +2015,51 @@ def tril(m, k=0):
...
+    >>> np.tri(3, 5, -1)
+    array([[0.,  0.,  0.,  0.,  0.],
+           [1.,  0.,  0.,  0.,  0.],
+           [1.,  1.,  0.,  0.,  0.]])
+    """
+    # if dtype is None:
+    #     dtype = _np.float32
+    # if M is None:
+    #     M = N
+    # return _npi.tri(N=N, M=M, k=k, dtype=dtype)
```
Comment (hzfan) [inline python/mxnet/ndarray/numpy/_op.py]: Dead code?

**Comment by hzfan** [comment]
FFI LGTM

File: python/mxnet/ndarray/numpy/_op.py [hunk 3]
```diff
@@ -2054,6 +2054,149 @@ def tril(m, k=0):
...
+    array([[1, 1, 1, 0, 0],
+           [1, 1, 1, 1, 0],
+           [1, 1, 1, 1, 1]])
+
+    >>> np.tri(3, 5, -1)
+    array([[0.,  0.,  0.,  0.,  0.],
+           [1.,  0.,  0.,  0.,  0.],
+           [1.,  1.,  0.,  0.,  0.]])
+    """
+    return _api_internal.tri(N, M, k, dtype, ctx)
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py:2094]:
```python
if ctx is None:
    ctx = current_context()
return _api_internal.tri(N, M, k, dtype, ctx)
```

Comment (haojin2) [reply python/mxnet/ndarray/numpy/_op.py:2094]: same for symbol

**Comment by Yiyan66** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by Yiyan66** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by haojin2** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by Yiyan66** [comment]
@mxnet-bot run ci [windows-gpu]

File: src/api/operator/numpy/np_tri_op.cc [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                String2MXNetTypeWithBool(args[3].operator std::string());
+  if (args[4].type_code() != kNull) {
+    attrs.dict["ctx"] = args[4].operator std::string();
+  }
+
+  attrs.parsed = param;
+  attrs.op = op;
+  SetAttrDict<op::TriParam>(&attrs);
+
+  int num_outputs = 1;
```
Comment (hzfan) [inline src/api/operator/numpy/np_tri_op.cc:53]: num_outputs should be 0. It means the number of outputs provided by the user, not the number of outputs produced by the op.

**Pull request merged**